src/sigchain/sigchain.iced

```diff
@@ -36,47 +53,62 @@ exports.ParsedKeys = ParsedKeys = class ParsedKeys
         # We also merge userids to help prove key ownership.
         existing.merge_everything key_manager
       else
-        kids_to_key_managers[kid_str] = key_manager
+        @_kids_to_merged_pgp_key_managers[kid_str] = key_manager
+      await key_manager.pgp_full_hash {}, esc defer hash
+      (@_kids_to_pgp_key_managers_by_hash[kid_str] or= {})[hash] = key_manager
+    else
+      @_kids_to_nacl_keys[kid_str] = key_manager
+    cb()
 
-      default_eldest_kid_for_testing or= kid_str
-    parsed_keys = new ParsedKeys {kids_to_key_managers}
-    cb null, parsed_keys, default_eldest_kid_for_testing
+  get_merged_pgp_key_manager : (kid) ->
+    @_kids_to_merged_pgp_key_managers[kid]
 
-  constructor : ({kids_to_key_managers}) ->
-    # Callers should use this class to get KeyManagers for KIDs they already
-    # have, but callers MUST NOT iterate over the set of KIDs here as though it
-    # were the valid set for a given user. The set of KIDs is *untrusted*
-    # because it comes from the server. We keep the map private to prevent that
-    # mistake.
-    @_kids_to_key_managers = kids_to_key_managers
+  get_pgp_key_manager_with_hash : (kid, hash) ->
+    @_kids_to_pgp_key_managers_by_hash[kid]?[hash]
 
-  get_key_manager : (kid) ->
-    @_kids_to_key_managers[kid]
+  get_nacl_key_manager : (kid) ->
+    @_kids_to_nacl_keys[kid]
+
+class KeyState
```

Comment:
Let's describe what this class is doing. (Sorry for all the comments about comments.)